File: Objective-C/admob/FullScreenNativeExample/FullScreenNativeExample/MainViewController.m
// Copyright 2024 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#import "MainViewController.h"

#import <GoogleMobileAds/GoogleMobileAds.h>

#import "GoogleMobileAdsConsentManager.h"
#import "NativeAdViewController.h"

@interface MainViewController () <GADNativeAdLoaderDelegate>

// You must keep a strong reference to the GADAdLoader during the ad loading process.
@property(strong, nonatomic) GADAdLoader *adLoader;

@property(weak, nonatomic) IBOutlet UIBarButtonItem *privacySettingsButton;
@property(weak, nonatomic) IBOutlet UIBarButtonItem *adInspectorButton;
@property(weak, nonatomic) IBOutlet UIButton *loadAdButton;
@property(weak, nonatomic) IBOutlet UIButton *showAdButton;

@property(strong, nonatomic) GADNativeAd *nativeAd;

@end

@implementation MainViewController

- (void)viewDidLoad {
  [super viewDidLoad];

  __weak __typeof__(self) weakSelf = self;
  [GoogleMobileAdsConsentManager.sharedInstance
      gatherConsentFromConsentPresentationViewController:self
                                consentGatheringComplete:^(NSError *_Nullable consentError) {
                                  if (consentError) {
                                    // Consent gathering failed.
                                    NSLog(@"Error: %@", consentError.localizedDescription);
                                  }

                                  __strong __typeof__(self) strongSelf = weakSelf;
                                  if (!strongSelf) {
                                    return;
                                  }

                                  if (GoogleMobileAdsConsentManager.sharedInstance.canRequestAds) {
                                    [strongSelf startGoogleMobileAdsSDK];
                                  }

                                  [strongSelf.loadAdButton
                                      setEnabled:GoogleMobileAdsConsentManager.sharedInstance
                                                     .canRequestAds];

                                  strongSelf.privacySettingsButton.enabled =
                                      GoogleMobileAdsConsentManager.sharedInstance
                                          .isPrivacyOptionsRequired;
                                }];

  // This sample attempts to load ads using consent obtained in the previous session.
  if (GoogleMobileAdsConsentManager.sharedInstance.canRequestAds) {
    [self startGoogleMobileAdsSDK];
  }
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
  // Pass the native ad to the full-screen native ad view controller.
  if ([segue.destinationViewController isKindOfClass:[NativeAdViewController class]]) {
    NativeAdViewController *viewController =
        (NativeAdViewController *)segue.destinationViewController;
    viewController.nativeAd = self.nativeAd;
  }
}

- (void)startGoogleMobileAdsSDK {
  static dispatch_once_t onceToken;
  dispatch_once(&onceToken, ^{
    // Initialize the Google Mobile Ads SDK.
    [GADMobileAds.sharedInstance startWithCompletionHandler:nil];

    // Configure the ad loader.
    GADNativeAdMediaAdLoaderOptions *mediaLoaderOptions =
        [[GADNativeAdMediaAdLoaderOptions alloc] init];
    mediaLoaderOptions.mediaAspectRatio = GADMediaAspectRatioPortrait;
    self.adLoader = [[GADAdLoader alloc] initWithAdUnitID:@"ca-app-pub-3940256099942544/5406332512"
                                       rootViewController:nil
                                                  adTypes:@[ GADAdLoaderAdTypeNative ]
                                                  options:@[ mediaLoaderOptions ]];
    self.adLoader.delegate = self;
  });
}

- (IBAction)privacySettingsButtonPressed:(UIBarButtonItem *)sender {
  [GoogleMobileAdsConsentManager.sharedInstance
      presentPrivacyOptionsFormFromViewController:self
                                completionHandler:^(NSError *_Nullable formError) {
                                  if (formError) {
                                    UIAlertController *alertController = [UIAlertController
                                        alertControllerWithTitle:formError.localizedDescription
                                                         message:@"Please try again later."
                                                  preferredStyle:UIAlertControllerStyleAlert];
                                    UIAlertAction *defaultAction =
                                        [UIAlertAction actionWithTitle:@"OK"
                                                                 style:UIAlertActionStyleCancel
                                                               handler:nil];

                                    [alertController addAction:defaultAction];
                                    [self presentViewController:alertController
                                                       animated:YES
                                                     completion:nil];
                                  }
                                }];
}

- (IBAction)adInspectorTapped:(UIBarButtonItem *)sender {
  [GADMobileAds.sharedInstance
      presentAdInspectorFromViewController:self
                         completionHandler:^(NSError *_Nullable error) {
                           if (error) {
                             UIAlertController *alertController = [UIAlertController
                                 alertControllerWithTitle:error.localizedDescription
                                                  message:@"Please try again later."
                                           preferredStyle:UIAlertControllerStyleAlert];
                             UIAlertAction *defaultAction =
                                 [UIAlertAction actionWithTitle:@"OK"
                                                          style:UIAlertActionStyleCancel
                                                        handler:^(UIAlertAction *action){
                                                        }];

                             [alertController addAction:defaultAction];
                             [self presentViewController:alertController
                                                animated:YES
                                              completion:nil];
                           }
                         }];
}

- (IBAction)loadAdButtonPressed:(UIButton *)sender {
  self.loadAdButton.enabled = NO;
  self.showAdButton.enabled = NO;

  [self.adLoader loadRequest:[GADRequest request]];
}

- (IBAction)showAdButtonPressed:(UIButton *)sender {
  self.loadAdButton.enabled = YES;
  self.showAdButton.enabled = NO;
}

#pragma mark GADAdLoaderDelegate implementation

- (void)adLoader:(GADAdLoader *)adLoader didReceiveNativeAd:(GADNativeAd *)nativeAd {
  self.nativeAd = nativeAd;
  self.showAdButton.enabled = YES;
}

- (void)adLoader:(GADAdLoader *)adLoader didFailToReceiveAdWithError:(NSError *)error {
  NSLog(@"%@ failed with error: %@", adLoader, error);
  self.loadAdButton.enabled = YES;
}

@end

File: Swift/advanced/APIDemo/APIDemo/MainViewController.swift
//
//  Copyright 2016-2023 Google LLC
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//

import UIKit

class MainViewController: UITableViewController {

  /// API Demo names.
  var apiDemoNames: [String]!

  /// Segue identifiers.
  var identifiers: [String]!

  override func viewDidLoad() {
    super.viewDidLoad()
    apiDemoNames = [
      "AdMob - Ad Delegate", "AdMob - Ad Targeting", "AdMob - Banner Sizes",
      "AdMob - Native Custom Mute This Ad", "AdMob - Ad Preloading",
      "AdManager - PPID", "AdManager - Custom Targeting", "AdManager - Category Exclusions",
      "AdManager - Multiple Ad Sizes", "AdManager - App Events", "AdManager - Fluid Ad Size",
      "AdManager - Custom Video Controls", "Collapsible Banner Ad",
    ]
    identifiers = [
      "adDelegateSegue", "adTargetingSegue", "bannerSizesSegue", "customMuteSegue",
      "adPreloadingSegue", "PPIDSegue", "customTargetingSegue", "categoryExclusionsSegue",
      "multipleAdSizesSegue", "appEventsSegue", "fluidAdSizeSegue",
      "customControlsSegue", "collapsibleSegue",
    ]
  }

  override func viewWillAppear(_ animated: Bool) {
    super.viewWillAppear(animated)
    if let indexPathSelected = tableView.indexPathForSelectedRow {
      tableView.deselectRow(at: indexPathSelected, animated: animated)
    }
  }

  // MARK: - Table View

  override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
    return apiDemoNames.count
  }

  override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath)
    -> UITableViewCell
  {
    let cell = tableView.dequeueReusableCell(withIdentifier: "DemoCell")!
    cell.textLabel!.text = apiDemoNames[indexPath.row]

    return cell
  }

  override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
    let row = indexPath.row
    if row < identifiers.count {
      performSegue(withIdentifier: identifiers[row], sender: self)
    }
  }

}
